Gems/Multiplayer/Code/Tests/MockInterfaces.h

```diff
@@ -0,0 +1,166 @@
+        MOCK_METHOD0(ClearAllEntities, void());
+        MOCK_METHOD1(AddEntityMarkedDirtyHandler, void(AZ::Event<>::Handler&));
+        MOCK_METHOD1(AddEntityNotifyChangesHandler, void(AZ::Event<>::Handler&));
+        MOCK_METHOD1(AddEntityExitDomainHandler, void(AZ::Event<const Multiplayer::ConstNetworkEntityHandle&>::Handler&));
+        MOCK_METHOD1(AddControllersActivatedHandler, void(AZ::Event<const Multiplayer::ConstNetworkEntityHandle&, Multiplayer::
+            EntityIsMigrating>::Handler&));
+        MOCK_METHOD1(AddControllersDeactivatedHandler, void(AZ::Event<const Multiplayer::ConstNetworkEntityHandle&, Multiplayer::
+            EntityIsMigrating>::Handler&));
+        MOCK_METHOD0(NotifyEntitiesDirtied, void());
+        MOCK_METHOD0(NotifyEntitiesChanged, void());
+        MOCK_METHOD2(NotifyControllersActivated, void(const Multiplayer::ConstNetworkEntityHandle&, Multiplayer::EntityIsMigrating));
+        MOCK_METHOD2(NotifyControllersDeactivated, void(const Multiplayer::ConstNetworkEntityHandle&, Multiplayer::EntityIsMigrating));
+        MOCK_METHOD1(HandleLocalRpcMessage, void(Multiplayer::NetworkEntityRpcMessage&));
+    };
+
+    class MockConnectionListener : public AzNetworking::IConnectionListener
+    {
+    public:
+        MOCK_METHOD3(ValidateConnect, ConnectResult(const IpAddress&, const IPacketHeader&, ISerializer&));
+        MOCK_METHOD1(OnConnect, void(IConnection*));
+        MOCK_METHOD3(OnPacketReceived, PacketDispatchResult (IConnection*, const IPacketHeader&, ISerializer&));
+        MOCK_METHOD2(OnPacketLost, void(IConnection*, PacketId));
+        MOCK_METHOD3(OnDisconnect, void(IConnection*, DisconnectReason, TerminationEndpoint));
+    };
+
+    class MockTime : public AZ::ITime
+    {
+    public:
+        MOCK_CONST_METHOD0(GetElapsedTimeMs, AZ::TimeMs());
+    };
+
+    class MockNetworkTime : public Multiplayer::INetworkTime
+    {
+    public:
+        MOCK_METHOD2(ForceSetTime, void (Multiplayer::HostFrameId, AZ::TimeMs));
+        MOCK_CONST_METHOD0(GetHostBlendFactor, float ());
+        MOCK_METHOD1(AlterBlendFactor, void (float));
+        MOCK_CONST_METHOD0(IsTimeRewound, bool());
+        MOCK_CONST_METHOD0(GetHostFrameId, Multiplayer::HostFrameId());
+        MOCK_CONST_METHOD0(GetUnalteredHostFrameId, Multiplayer::HostFrameId());
+        MOCK_METHOD0(IncrementHostFrameId, void());
+        MOCK_CONST_METHOD0(GetHostTimeMs, AZ::TimeMs());
+        MOCK_CONST_METHOD0(GetRewindingConnectionId, AzNetworking::ConnectionId());
+        MOCK_CONST_METHOD1(GetHostFrameIdForRewindingConnection, Multiplayer::HostFrameId(AzNetworking::ConnectionId));
+        MOCK_METHOD3(AlterTime, void(Multiplayer::HostFrameId, AZ::TimeMs, AzNetworking::ConnectionId));
+        MOCK_METHOD1(SyncEntitiesToRewindState, void(const AZ::Aabb&));
+        MOCK_METHOD0(ClearRewoundEntities, void());
+    };
+
+    class MockComponentApplicationRequests : public AZ::ComponentApplicationRequests
```

Comment:
I really dislike all these mocks of core interfaces getting duplicated everywhere, it's making the coupling to these interfaces far more complicated than necessary. As an example, updating or adding a new interface now requires editing over a dozen different mocks across almost the same number of GEMs.
This is completely inverted from what *should* be happening, which is we should be providing these empty mocks as part of the testing framework that defines the interface itself, so AzCore.Tests or AzFramework.Tests or AzNetworking.Tests, and for specialized testing use-cases that require non-stub functionality then they can opt to not use the default provided mock and instead implement their own.
The current pattern used elsewhere and duplicated here is not scaling at all.

Reply:
I agree. It has been an absolute pain to set these up. That is why I have to start to think how to improve these.